pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java

```diff
@@ -43,17 +44,79 @@
  */
 @SuppressWarnings("unused")
 public class ServiceStatus {
+  public static final String STATUS_DESCRIPTION_NONE = "None";
+  public static final String STATUS_DESCRIPTION_INIT = "Init";
+  public static final String STATUS_DESCRIPTION_STARTED = "Started";
+  public static final String STATUS_DESCRIPTION_NO_HELIX_STATE = "Helix state does not exist";
   private static final Logger LOGGER = LoggerFactory.getLogger(ServiceStatus.class);
+  private static final int MAX_RESOURCE_NAMES_TO_LOG = 5;
+  private static final Map<String, ServiceStatusCallback> serviceStatusCallbackMap = new ConcurrentHashMap<>();
+  private static final ServiceStatusCallback serviceStatusCallback =
+      new MapBasedMultipleCallbackServiceStatusCallback(serviceStatusCallbackMap);
 
-  public enum Status {
-    STARTING, GOOD, BAD
+  public static void setServiceStatusCallback(String name, ServiceStatusCallback serviceStatusCallback) {
+    ServiceStatus.serviceStatusCallbackMap.put(name, serviceStatusCallback);
   }
 
-  public static final String STATUS_DESCRIPTION_NONE = "None";
-  public static final String STATUS_DESCRIPTION_INIT = "Init";
-  public static final String STATUS_DESCRIPTION_NO_HELIX_STATE = "Helix state does not exist";
+  public static void removeServiceStatusCallback(String name) {
+    ServiceStatus.serviceStatusCallbackMap.remove(name);
+  }
 
-  private static final int MAX_RESOURCE_NAMES_TO_LOG = 5;
+  public static Status getServiceStatus() {
+    return getServiceStatus(serviceStatusCallback);
+  }
+
+  public static Status getServiceStatus(String name) {
+    if (serviceStatusCallbackMap.containsKey(name)) {
+      return getServiceStatus(serviceStatusCallbackMap.get(name));
+    } else {
+      return Status.NOT_STARTED;
```

Comment:
We add it to the map when we call start() and remove it from the map when we call stop(). So,  the status here can either be NOT_STARTED, or SHUTTING_DOWN. 

I suggest you add it to the map in the constructor, and return SHUTTING_DOWN here once you remove from the map.

Reply:
We  have status of `NOT_STARTED`, `STARTING`, `GOOD`, `BAD` for now.
I feel it's ok to say `NOT_STARTED` after we stop an instance, as it's still possible to start a new instance later on.

Reply:
But then the shut down sequence is not instantaneous. You call shutdown and then wait for it to shutdown. During that time, the status should be SHUTTING_DOWN. And then you can change it to NOT_STARTED when the shutdown is complete.

Reply:
Got it. Added `SHUTTING_DOWN` status.
My feeling is that during shutting down phase, the `getServiceStatus(serviceStatusCallbackMap.get(name))` should return `SHUTTING_DOWN`.

Once server is completely shutdown, it will be removed from the `serviceStatusCallbackMap` so the status will be `NOT_STARTED`